src/main/scala/persistence/MemStorage.scala

```diff
@@ -22,35 +23,28 @@ class MemStorage extends AuctionService with InputValidator {
 
   override def createAuction(data: AuctionData): VNel[AuctionId] = {
     def _createAuction(data: String): AuctionId = {
-      val newId = UUID.randomUUID.toString
+      val newId = newUUIDString
       val auction = Auction(newId, data)
       auctions :+= auction
       newId
     }
 
     Apply[VNel].apply(
-      validData(data).toSuccessNel(auctionDataErrorMsg)
+      validData(data, auctionDataErrorMsg)
     ) {
       _ => _createAuction(data)
     }
   }
 
-  override def getAuction(id: AuctionId): Option[Auction] = auctions.find( _.id == id)
+  override def getAuction(id: AuctionId): VNel[Auction] = auctions.find( _.id == id).toSuccessNel(auctionNotFoundErrorMsg)
 
   override def addLot(auctionId: AuctionId, data: LotData): VNel[LotId] = {
-    def _addLot(auctionId: AuctionId, data: LotData): LotId = {
-      val newId = UUID.randomUUID.toString
-      val lot = Lot(newId, auctionId, data)
-      lots :+= lot
-      newId
-    }
-
-    (
-      validUUIDString(auctionId).toSuccessNel(auctionIdErrorMsg) |@|
-      getAuction(auctionId).toSuccessNel(auctionNotFoundErrorMsg) |@|
-      validData(data).toSuccessNel(lotDataErrorMsg)
-    ) {
-      (_, _, _) => _addLot(auctionId, data)
+    newLot(auctionId, data) match {
```

Comment:
As you're not doing anything with `Failure`, you can make it shorter by using `map`:
```
newLot(auctionId, data) map { lot =>
  lots :+= lot
  lot.id
}
```

Reply:
Perfect. I was looking for something like that and it's obvious now that you wrote it ;)